source/grammar/qasm3.g4

```diff
@@ -276,34 +279,73 @@ logicalOperator
     | '||'
     ;
 
+shiftOperator
+    : '<<'
+    | '>>'
+    ;
+
+additiveOperator
+    : '+'
+    | '-'
+    ;
+
+multiplicativeOperator
+    : '*'
+    | '/'
+    | '%'
+    ;
+
+powerOperator
+    : '**'
+    ;
+
 expressionStatement
     : expression SEMICOLON
     ;
 
 expression
     // include terminator/unary as base cases to simplify parsing
     : expressionTerminator
-    | unaryExpression
     // expression hierarchy
-    | xOrExpression
-    | expression '|' xOrExpression
+    | booleanExpression
     ;
 
 /**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
   *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
   * Order (first to last evaluation):
     Terminator (including Parens),
     Unary Op,
+    Power
     Multiplicative
     Additive
     Bit Shift
     Bit And
     Exlusive Or (xOr)
     Bit Or
+    Comparison
+    Logical (&&, ||)
+    bool expr
+    membership
 **/
-xOrExpression
+booleanExpression
+    : membershipTest
+    | comparisonExpression
+    | booleanExpression logicalOperator comparisonExpression
+    ;
+
+comparisonExpression
+    : orExpression  // if (expression)
+    | comparisonExpression relationalOperator orExpression
```

Comment:
Again, the reason I used `expression` here is that for simple cases like expression terminator and unary, you don't need to go through the entire hierarchy. Makes the resulting parse tree far clearer and is still accurate.